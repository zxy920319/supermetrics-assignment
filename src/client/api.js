import { get, post, getPaginated } from "@/client"

const getToken = async () => {
  try {
    const { data } = await post(
      "/register",
      {
        client_id: import.meta.env.VITE_CLIENT_ID,
        email: import.meta.env.VITE_EMAIL,
        name: import.meta.env.VITE_NAME,
      },
      {},
      true
    )
    const { sl_token } = data
    return sl_token
  } catch (error) {
    console.log(error)
  }
}

const getPosts = async (params) => {
  try {
    const { data } = await get("/posts", { params })
    const { posts } = data
    return posts
  } catch (error) {
    console.log(error)
  }
}

const getAllPosts = async (params) => {
  try {
    const posts = await getPaginated("/posts", params)
    return posts
  } catch (error) {
    console.log(error)
  }
}

export { getToken, getPosts, getAllPosts }
