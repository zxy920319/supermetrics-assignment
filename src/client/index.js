import axios from "axios"

const client = axios.create({ baseURL: import.meta.env.VITE_APP_API_URL })

const get = async (url, config) => {
  try {
    const response = await client.get(url, config)
    return response.data
  } catch (error) {
    throw new Error(error)
  }
}

const post = async (url, data, config) => {
  try {
    const response = await client.post(url, data, config)
    return response.data
  } catch (error) {
    throw new Error(error)
  }
}

const getToken = async () => {
  try {
    const { data } = await post("/register", {
      client_id: import.meta.env.VITE_CLIENT_ID,
      email: import.meta.env.VITE_EMAIL,
      name: import.meta.env.VITE_NAME,
    })
    const { sl_token } = data
    return sl_token
  } catch (error) {
    throw new Error(error)
  }
}

const getPosts = async (params) => {
  try {
    const { data } = await get("/posts", { params })
    const { posts } = data
    return posts
  } catch (error) {
    throw new Error(error)
  }
}

export { get, post, getToken, getPosts }
