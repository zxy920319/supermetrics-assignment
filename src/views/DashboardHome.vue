<template>
  <DashboardContent>
    <div class="min-h-screen bg-white rounded-xl shadow-md p-4">
      <ATable :data-source="posts" :columns="columns" :pagination="pagination" :loading="loading" @change="handleChange">
        <template #bodyCell="{ column, record }">
          <template v-if="column.key === 'created_time'">
            {{ formatDateTime(record.created_time) }}
          </template>
        </template>
      </ATable>
    </div>
  </DashboardContent>
</template>

<script setup>
import { useAuthStore } from "@/stores/auth"
import { getPosts } from "@/client/api"
import { reactive, ref } from "@vue/reactivity"
import { inject, onMounted } from "@vue/runtime-core"
import DashboardContent from "@/components/DashboardContent.vue"
import { Table as ATable } from "ant-design-vue"
import "ant-design-vue/es/table/style/css"

const formatDateTime = inject("formatDateTime")
const store = useAuthStore()
const posts = ref([])
const loading = ref(true)
const pagination = reactive({
  current: 1,
  showSizeChanger: false,
  showLessItems: true,
  pageSize: 10,
})
const columns = [
  { title: "Name", dataIndex: "from_name", key: "name", width: 150 },
  { title: "Comment", dataIndex: "message", key: "message" },
  { title: "Commented on", dataIndex: "created_time", key: "created_time", width: 200, fixed: "right" },
]

const fetchPosts = async (page) => {
  const params = {
    sl_token: store.token,
    page,
  }
  const response = await getPosts(params)
  return response
}

const handleChange = async ({ current }) => {
  if (current !== pagination.current) {
    pagination.current = current
    posts.value = await fetchPosts(current)
  }
}

onMounted(async () => {
  posts.value = await fetchPosts(pagination.current)
  loading.value = false
})
</script>
