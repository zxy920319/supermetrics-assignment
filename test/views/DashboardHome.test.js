import DashboardHome from "@/views/DashboardHome.vue"
import { describe, expect, it, vi } from "vitest"
import { flushPromises, mount } from "@vue/test-utils"
import { createTestingPinia } from "@pinia/testing"
import { getPosts } from "@/client/api"

vi.mock("@/client/api", () => ({
  getPosts: vi.fn().mockResolvedValue([]),
}))
Object.defineProperty(window, "matchMedia", {
  writable: true,
  value: vi.fn().mockImplementation((query) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: vi.fn(), // deprecated
    removeListener: vi.fn(), // deprecated
    addEventListener: vi.fn(),
    removeEventListener: vi.fn(),
    dispatchEvent: vi.fn(),
  })),
})

describe("DashboardHome", async () => {
  const render = () => {
    return mount(DashboardHome, {
      global: {
        plugins: [createTestingPinia({ initialState: { auth: { token: "token" } }, createSpy: vi.fn })],
      },
    })
  }

  it("calls fetchPosts on mounted", async () => {
    const wrapper = await render()
    await flushPromises()
    const params = { sl_token: "token", page: 1 }
    expect(getPosts).toHaveBeenCalledWith(params)
    expect(wrapper.html()).toMatchSnapshot()
  })
})
