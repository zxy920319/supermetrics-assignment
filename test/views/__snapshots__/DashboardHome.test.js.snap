// Vitest Snapshot v1

exports[`DashboardHome > calls fetchPosts on mounted 1`] = `
"<div class=\\"min-h-screen bg-gray-200 pb-20\\">
  <div class=\\"bg-[#03A9F4] px-3 md:px-8 h-56\\"></div>
  <div class=\\"px-20 -mt-24\\">
    <div class=\\"container mx-auto max-w-full\\">
      <div class=\\"min-h-screen bg-white rounded-xl shadow-md p-4\\">
        <div class=\\"ant-table-wrapper\\">
          <div class=\\"ant-spin-nested-loading\\">
            <!---->
            <div class=\\"ant-spin-container\\">
              <div class=\\"ant-table ant-table-has-fix-right ant-table-empty\\">
                <!---->
                <div class=\\"ant-table-container\\">
                  <div class=\\"ant-table-content\\">
                    <table style=\\"table-layout: auto;\\">
                      <colgroup>
                        <col style=\\"width: 150px;\\">
                        <col>
                        <col style=\\"width: 200px;\\">
                      </colgroup>
                      <thead class=\\"ant-table-thead\\">
                        <tr>
                          <th class=\\"ant-table-cell\\" colstart=\\"0\\" colend=\\"0\\">
                            <!---->Name
                            <!---->
                          </th>
                          <th class=\\"ant-table-cell\\" colstart=\\"1\\" colend=\\"1\\">
                            <!---->Comment
                            <!---->
                          </th>
                          <th class=\\"ant-table-cell ant-table-cell-fix-right ant-table-cell-fix-right-first\\" colstart=\\"2\\" colend=\\"2\\" style=\\"position: sticky; right: 0px;\\">
                            <!---->Commented on
                            <!---->
                          </th>
                        </tr>
                      </thead>
                      <tbody class=\\"ant-table-tbody\\">
                        <!---->
                        <tr class=\\"ant-table-placeholder\\">
                          <td colspan=\\"3\\" class=\\"ant-table-cell\\">
                            <!---->
                            <div class=\\"ant-empty ant-empty-normal\\">
                              <div class=\\"ant-empty-image\\"><svg class=\\"ant-empty-img-simple\\" width=\\"64\\" height=\\"41\\" viewBox=\\"0 0 64 41\\">
                                  <g transform=\\"translate(0 1)\\" fill=\\"none\\" fill-rule=\\"evenodd\\">
                                    <ellipse class=\\"ant-empty-img-simple-ellipse\\" fill=\\"#F5F5F5\\" cx=\\"32\\" cy=\\"33\\" rx=\\"32\\" ry=\\"7\\"></ellipse>
                                    <g class=\\"ant-empty-img-simple-g\\" fill-rule=\\"nonzero\\" stroke=\\"#D9D9D9\\">
                                      <path d=\\"M55 12.76L44.854 1.258C44.367.474 43.656 0 42.907 0H21.093c-.749 0-1.46.474-1.947 1.257L9 12.761V22h46v-9.24z\\"></path>
                                      <path d=\\"M41.613 15.931c0-1.605.994-2.93 2.227-2.931H55v18.137C55 33.26 53.68 35 52.05 35h-40.1C10.32 35 9 33.259 9 31.137V13h11.16c1.233 0 2.227 1.323 2.227 2.928v.022c0 1.605 1.005 2.901 2.237 2.901h14.752c1.232 0 2.237-1.308 2.237-2.913v-.007z\\" fill=\\"#FAFAFA\\" class=\\"ant-empty-img-simple-path\\"></path>
                                    </g>
                                  </g>
                                </svg></div>
                              <p class=\\"ant-empty-description\\">No Data</p>
                              <!---->
                            </div>
                            <!---->
                          </td>
                        </tr>
                      </tbody>
                      <!---->
                    </table>
                  </div>
                </div>
                <!---->
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>"
`;
